Source/AuthenticationInterceptor.swift

```diff
@@ -338,14 +338,16 @@ public class AuthenticationInterceptor<AuthenticatorType>: RequestInterceptor wh
         mutableState.refreshTimestamps.append(ProcessInfo.processInfo.systemUptime)
         mutableState.isRefreshing = true
 
-        authenticator.refresh(credential, for: session) { result in
-            self.$mutableState.write { mutableState in
-                switch result {
-                case let .success(credential):
-                    self.handleRefreshSuccess(credential, insideLock: &mutableState)
-
-                case let .failure(error):
-                    self.handleRefreshFailure(error, insideLock: &mutableState)
+        // Dispatch to queue to hop out of the lock in case authenticator.refresh is implemented synchronously.
+        queue.async {
+            self.authenticator.refresh(credential, for: session) { result in
+                self.$mutableState.write { mutableState in
+                    switch result {
+                    case let .success(credential):
+                        self.handleRefreshSuccess(credential, insideLock: &mutableState)
+                    case let .failure(error):
+                        self.handleRefreshFailure(error, insideLock: &mutableState)
+                    }
                 }
             }
         }
```

Comment:
I think this approach is fine. It's unfortunate that we need to `hop` for the rare case that someone didn't already hop. The only other thing I'd say we could consider is instead hopping in the completion rather than before the call. In the end though I don't think it matters either way. This will fix the issue and I don't have any better solution.

Reply:
Tests work either way. However, I am a bit concerned about executing user code while still locked, so I'll leave it as it is. Unfortunately I think the real fix to refactor with direct locking rather than the closure API, but we can do that later.